src/qli/java/com/lambdaherding/edi/qli/ch02/Calculate.java

```diff
@@ -0,0 +1,170 @@
+	
+	private JTextField display;
+	private StringBuilder number = new StringBuilder();
+	private Stack<Double> data = new Stack<>();
+
+	public Calculate() {
+		super( "My Calculator" );
+		setDefaultCloseOperation( JFrame.EXIT_ON_CLOSE );
+		
+		setLayout( new BorderLayout() );
+		
+		display = new JTextField();
+		display.setEditable( false );
+		add( display, BorderLayout.NORTH );
+		
+		JPanel buttonPanel = makeButtons();
+		add( buttonPanel, BorderLayout.CENTER );
+		
+		pack();
+	}
+
+	private JPanel makeButtons() {
+		JPanel pan = new JPanel();
+		pan.setLayout(new GridLayout(5, 5));
+
+		Arrays.asList( 
+				inputButton("7"), inputButton("8"), inputButton("9"),
+				binaryActionButton("/", (x, y) -> x / y), unaryActionButton("%", x -> x / 100),
+				inputButton("4"), inputButton("5"), inputButton("6"), 
+				binaryActionButton("*", (x, y) -> x * y), unaryActionButton("1/x", x -> 1 / x),
+				inputButton("1"), inputButton("2"), inputButton("3"), 
+				binaryActionButton("-", (x, y) -> x - y), unaryActionButton("+/-", x -> -x),
+				commaButton(), inputButton("0"), inputButton("."), 
+				binaryActionButton("+", (x, y) -> x + y), unaryActionButton("\u221a", x -> Math.sqrt(x)),
+				clearAll("ClearAll")
+				).forEach( b -> pan.add(b) );		
+		return pan;
+	}
+
+	public JButton jButton(String label, ActionListener l) {
+		JButton button = new JButton(label);
+		button.setActionCommand(label);
+		button.addActionListener(l);
+		return button;
+	}
+
+	public JButton clearAll(String label) {
+		return jButton(label, (evt) -> {
+			display.setText("");
+			data.clear();
+			number = new StringBuilder();
+		});
+	}
+	
+	public JButton inputButton(String label) {
+		return jButton(label, (evt) -> {
+			addToDisplay(evt.getActionCommand());
+			number.append(evt.getActionCommand());
+		});
+	}
+
+	public JButton commaButton() {
+		return jButton(",", evt -> {
+			if (doParse()) {
+				addToDisplay(",");
+			}
+		});
+	}
+
+	private JButton unaryActionButton(String label, UnaryOperator<Double> unaryOp) {
+		return jButton(label, evt -> {
+			if (number.length() > 0) {
+				if (!doParse()) {
+					return;
+				}
+			}
+
+			if (data.empty()) {
+				JOptionPane.showMessageDialog(null, "No numbers on stack",
+						"Empty stack", JOptionPane.ERROR_MESSAGE);
+			} else {
+				double result = unaryOp.apply(data.pop());
+				data.push(result);
+				deleteLastNumber();
+				addToDisplay(result + ",");
+			}
+		});
+	}
+	
+	private JButton binaryActionButton(String label, BinaryOperator<Double> binaryOp) {
+		return jButton(label, evt -> {
+			if (number.length() > 0) {
+				if (!doParse()) {
+					return;
+				}
+			}
+
+			if (data.size() < 2) {
+				JOptionPane.showMessageDialog(null, "Not enough numbers on stack", "Stack too small", JOptionPane.ERROR_MESSAGE);
+			} else {
+				double result = binaryOp.apply(data.pop(), data.pop());
+				data.push(result);
+				deleteLastNumber();
+				deleteLastNumber();
+				addToDisplay(result + ",");
+			}
+		});
+	}
```

Comment:
You went for a similar approach to me here, I was hoping I could somehow merge unaryActionButton and binaryActionButton and deal with variable numbers of input arguments but I couldn't find a way